Logical operations over BuiltinByteString

kozross opened the pull request:
This CIP describes several operations over bits, including AND, OR and XOR, as well as indexing, setting, and clearing, operating on `BuiltinByteString`. Additionally, it provides a 'replication' operation to produce a `BuiltinByteString` of a given length full of identical bytes.

rphair reviewed (commented):
thanks @kozross, looks great - please keep us updated in PR comments about how the path-to-active progresses. 🙏 

@michaelpj @zliu41 can you review this?  I've added to the agenda for introduction at the next CIP Editor meeting: https://hackmd.io/@cip-editors/88

Ryun1 commented:
@michaelpj @zliu41 comment would be appreciated

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+the value at index $i$ of $b$ (and analogously, $b_r$); see the [section on the bit
+indexing scheme](#bit-indexing-scheme) for the exact specification of this.
+
+If the change list argument is empty, the result is $b$. If the change list
```
I don't think it would be a problem to define the algorithm by recursion. Given the audience for this kind of document, I think that would probably be a clearer way of communicating it!

kozross replied:
I was kind of going for that exactly, but clearly that didn't work so well. Can you suggest a better way I could have phrased this?

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+We describe precisely the operations we intend to implement, and their
+semantics. These operations will have the following signatures:
+
+* `builtinLogicalAnd :: BuiltinBool -> BuiltinByteString -> BuiltinByteString ->
```
The names are odd. We don't normally prefix all operations with "builtin", we do sometimes indicate something about the type. So if anything I would have thought we might want "bytestringLogicalAnd" or something.

kozross replied:
I'm open to different names. I ended up going with `bitwiseLogicalAnd` in the PR to Plutus Core.

kozross replied:
I've renamed the operations as follows:

* `builtinLogicalAnd` -> `bitwiseLogicalAnd` (similar for OR, XOR and complement)
* `builtinReadBit` -> `readBit`
* `builtinSetBits` -> `writeBits`
* `builtinReplicate` -> `replicateByteString`

This matches the choices I made when defining these operations in the [Plutus PR](https://github.com/IntersectMBO/plutus/pull/5970) designed to implement this CIP.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+  byte; and
+* What `BuiltinInteger` was passed as the byte argument.
+
+Otherwise, let $b$ be the result of `builtinReplicate`, and let $b\\{i\\}$ be the
```
```suggestion
Otherwise, let $b$ be the result of `builtinReplicate`, and let $b\\[i\\]$ be the
```
worth checking other uses for consistency. Shame we don't have macros...

kozross replied:
I deliberately used `{}` here, as I'm indexing bytes, not bits, which is what `[]` is used for.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+scheme](#bit-indexing-scheme). We have:
+
+* The length (in bytes) of $b$ is $n$; and
+* For all $i \in 0, 1, \ldots, n - 1$, $b\\{i\\} = w$.
```
```suggestion
* For all $i \in 0, 1, \ldots, n - 1$, $b\\[i\\] = w$.
```

kozross replied:
Same here.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+values at bit indexes $0$ through $7$, and treat them as their corresponding
+place value in a summation, we should obtain the same answer as indexing
+whichever byte those bits come from. We call this notion _conversion agreement_,
+due to its relation to [CIP-121][conversion-cip], which allows us to
```
CIP-121 allows for different endianness choices, does that not also matter here?

kozross replied:
It matters for one of the choices, namely the big-endian one. I should state this more clearly.

kozross replied:
It matters for one of the choices, namely the big-endian one. I should state this more clearly.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+all: the only possible signature such an operation would have is
+`[BuiltinByteString] -> [BuiltinByteString]`, which in effect would be a
+specialized form of mapping. Even the _general_ form of mapping is not
+considered suitable as a primitive operation in Plutus Core! 
```
Aside, but it's not that it's not considered suitable, but rather that we can't define it since we can't define higher-order builtin functions right now.

kozross replied:
Fair point, I'll amend.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+potentially have been implemented differently based on existing work: most
+notably, our choice of bit indexing scheme, the padding-versus-truncation
+semantics, and the multiplicitous definition of bit modification. Among existing
+work, a particularly important example is [CIP-58][cip-58], which makes
```
One thing I'm missing is a discussion of the things that are not included from CIP-58. It's unclear to me whether the intention is simply for this to be another tranche of that work, or whether we've decided that some of the CIP-58 operations are less useful.

kozross replied:
This is meant to be a continuation of CIP-121, which was meant to break 58 into manageable chunks, as well as to address some of the shortcomings of the original work. I should also make this more clear.

michaelpj commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+We also note that conversion agreement matters for Case 2, in the wider context
+of the interaction between [CIP-121 primitives][conversion-cip] and logical
+XOR. The Argon2 family of hashes use certain inputs (which happen to be numbers)
+both as numbers and also as blocks of binary, which means that any inconsistency
```
I think this is a good example, and it leaves me uncertain how this would help us. Is the idea that:
- We somehow make 0-cost conversions from bytestrings to integers to facilitate this case
- Implementors of such algorithms implement their own integer operations on bytestrings based on the given representation
?

kozross replied:
This means that we don't have to worry about this disagreement in the first place. People can convert integers to bytestrings, fiddle with their bits, then convert back, and expect 'the right results' at the other end.

michaelpj replied:
I see, so it's something like my second bullet point: you can convert to a bytestring, do some fiddling that you _expect_ to have consequences for the encoded integer, convert back, and be correct.

kozross replied:
Yes, exactly. I feel this is an essential feature.

michaelpj replied:
I feel like I got slightly lost in the exposition, but are we definitely sure that nobody will want conversion agreement with a _different_ arrangement? For example, with the other endianness.

kozross replied:
It's not possible to have all of the following:

* Consecutive bit indexes stay consecutive
* Conversion agreement with little endian ordering
* My 1-byte example to make sense

Given that, I had to sacrifice something, and I feel sacrificing conversion agreement with little-endian ordering was the right choice. Do you feel I should state this more explicitly?

michaelpj replied:
Yes, that would be helpful. Naively I might have thought that conversion agreement with little-endian ordering (if that's important for some reason!) might be _more_ important than some of the other things. Also, I just found it a bit confusing that you referred to conversion-agreement _unqualified_, when it's conversion agreement with a particular encoding scheme that's chosen as important for... some reason.

kozross replied:
Yeah, upon re-reading, I agree, it could be more clear. Will add.

michaelpj reviewed (approved):
Good work as usual! I'm basically happy with this.

kozross commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+2$, and $3 - 2 - 1 = 0$. Within the byte $42$, this means we have
+$\left\lfloor\frac{42}{2^3}\right\rfloor\mod 2 \equiv 1$. Thus, $b[19] = 1$. 
+
+Put another way, our _byte_ indexes run 'the opposite way' to our _bit_ indexes.
```
It should be the same. I feel this explanation is a lot clearer though.

kozross commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1408 @@
+  BuiltinByteString`
+* `builtinReplicate :: BuiltinInteger -> BuiltinInteger -> BuiltinByteString`
+
+We assume the following costing:
```
Yes, but I will amend this to note both costings.

rphair commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1437 @@
+---
+CIP: ???
```
```suggestion
CIP: 122
```
Please also change directory name, and also the link in the first comment in this thread to point to that directory or its `README.md` file.

rphair commented on CIP-XXX/CIP-XXX.md:
```diff
@@ -0,0 +1,1437 @@
+Implementors: 
+    - Koz Ross <koz@mlabs.city>
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/?
```
```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/806
```

rphair reviewed (commented):
@kozross excellent response to this at the CIP Editors' Meeting on Tuesday of this week.  Editors (cc @Ryun1 @Crypto2099) will look to responses to points from @michaelpj's https://github.com/cardano-foundation/CIPs/pull/806#pullrequestreview-2057539210 to see what to do next.

re: my https://github.com/cardano-foundation/CIPs/pull/806#pullrequestreview-2037471583 about Path to Active, it looks good to me now... althrough if you think those bulleted items might progress intermittently I would suggest turning them into tickboxes.

kozross commented:
@rphair and @michaelpj - I've addressed all feedback from both of your reviews. Please let me know if something else needs doing; otherwise, I feel this should merge.

rphair commented on README.md:
```diff
@@ -101,6 +101,7 @@ CIP Editors meetings are public, recorded, and [published on Youtube](https://ww
 | 0110 | [Plutus v1 compatible script references](./CIP-0110) | Proposed |
 | 0114 | [CBOR Tags Registry](./CIP-0114) | Proposed |
 | 0115 | [CBOR tag definition - ED25519-BIP32 Keys](./CIP-0115) | Proposed |
+| 0122 | [Logical operations](./CIP-0122) | Proposed | 
```
```suggestion
```
@kozross I appreciate the initiative in adding this, but I'd only thrown this change into a CIP PR as per https://github.com/cardano-foundation/CIPs/pull/624#issuecomment-2092353644 because such an old proposal was being suddenly merged & legitimised.  Generally it helps editorially if editors themselves can submit one PR to do all these updates every 2 weeks so we don't end up with merge conflicts across branches from different submitters.

kozross replied:
No problem, can remove.

rphair reviewed (commented):
@kozross I can see that all the comments are responded & will advocate for merging this at next week's CIP meeting... I've put it on our [meeting agenda](https://hackmd.io/@cip-editors/89) for `Last Check`.  Normally CIPs stay an additional meeting cycle in the `Review` stage but assuming there are no further challenges from @michaelpj I will argue that due diligence has been done here regardless (cc @Ryun1 @Crypto2099).  I'll green-check this as soon as we handle the minor issue below.

@MicroProofs I think you had responded favourably to this at the last CIP meeting so please feel free to review with another set of eyes: to help the community review we'd normally get from longer exposure.

rphair reviewed (approved):
Looks good as of https://github.com/cardano-foundation/CIPs/pull/806/commits/ef3ecb9cf3f6500428abb16f55f1614d32491570 will invite other editors to confirm.

kozross commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+Our work relates to both [CIP-58][cip-58] and [CIP-121][cip-121]. Essentially,
+our goal with both this CIP and CIP-121 is to both break CIP-58 into more
+manageable (and reviewable) parts, and also address some of the design choices
+in CIP-58 that were not as good (or as clear) as they could have been. In this
```
CIP-58 is functionally deprecated _already_: the PR for its implementation was rejected for inclusion into Plutus. This work is fully designed to replace it. Therefore, there is no need to ask anyone's opinion: CIP-58 never got used in anything.

rphair replied:
thanks @kozross for clarification; cross-referencing https://github.com/cardano-foundation/CIPs/pull/825#issuecomment-2135947237.

kwxm commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+
+For the binary logical operations (that is, `bitwiseLogicalAnd`,
+`bitwiseLogicalOr` and `bitwiseLogicalXor`), the we have two choices of
+semantics when handling `BuiltinByteString` arguments of different lengths. We
```
... or you could fail.

kwxm commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+## Abstract
+
+We describe the semantics of a set of logical operations for Plutus
+`BuiltinByteString`s. Specifically, we provide descriptions for:
```
Now I'm wondering what "Plutus" really means in this document.  Our names are a bit confusing and strictly there's no such thing as Plutus: there are only (typed and untyped) Plutus Core and the Haskell thing called PlutusTx.  There's a suggestion that we should reduce confusion by reserving the name "Plutus" for the thing that runs on the chain (ie, untyped Plutus Core) and rename PlutusTx to something else.  I don't know when (or if) that's likely to happen though.

kozross replied:
This is a good point more generally: some clarity in what terms we use in CIPs would be good to have, similar to how something like [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) specifies what certain words mean exactly and standardizes their use.

kwxm reviewed (commented):
I was looking at this while I was reviewing the implementation and added a few comments about fairly superficial things, most of which don't need any response.  I didn't check things like the details of the semantics: I'll try to come back and review it more thoroughly later.

kozross commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+writeBits [0xFF] [(5, False)] => [0xDF]
+
+-- Larger indexes write backwards into the bytes from the end
+writeBits [0xF4, 0xFF] [(10, True)] => [0xF0, 0xFF]
```
You're absolutely right: looks like a copy-paste error. I'll have to fix these.

kozross commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+(at least not for _all_ inputs possible). While [CIP-58][cip-58] chose match
+semantics, we believe this was not the correct decision: we use Case 1 to
+justify the benefit of having other semantics described above available.
+
```
You're right that the assumption is straight from Haskell. It is in fact _so_ implicit I didn't even notice I'd done it! Definitely worth mentioning I think.

kozross commented on CIP-0122/CIP-0122.md:
```diff
@@ -0,0 +1,1488 @@
+
+### Bit setting
+
+`writeBits` in our description takes a change list argument, allowing
```
Thanks! I came at it from the point of memory use: if you're forced to copy something 10 times to change 10 bits, this is quite bad in a setting where you have limited resources.

kozross commented:
@Ryun1 - I believe I have addressed your concerns now.

@kwxm - thanks for catching two quite confusing mistakes on my part. I have now fixed these, and hopefully clarified my intent with both.

kozross commented:
@rphair - done. Sorry for the confusion on my part.
